godot-core/src/obj/dyn_gd.rs

```diff
@@ -137,6 +137,53 @@ use std::{fmt, ops};
 /// and it can query the dynamic type of the object. Based on that type, it can find the `impl Health` implementation matching the correct class.
 /// Behind the scenes, everything is wired up correctly so that you can restore the original `DynGd` even after it has passed through Godot.
 ///
+/// # Type inference
+///
+/// If class implements more than one `#[godot_dyn]` trait, type inference only works when the trait we dereference to explicitly declares
+/// `'static` as its supertrait.
+/// Otherwise, if only one `#[godot_dyn]` trait is implemented for a given class, the type can be properly inferred.
+///
+/// ```no_run
+/// # use godot::prelude::*;
+/// trait Health: 'static { /* ... */ }
+///
+/// // Exact equivalent to:
+/// trait OtherHealth
+/// where
+///     Self: 'static
+/// { /* ... */ }
+///
+/// trait NoInference { /* ... */ }
+///
+/// #[derive(GodotClass)]
+/// # #[class(init)]
+/// struct Monster { /* ... */ }
+///
+/// #[godot_dyn]
+/// impl Health for Monster { /* ... */ }
+///
+/// #[godot_dyn]
+/// impl NoInference for Monster { /* ... */ }
+///
+/// fn deal_damage(h: &mut dyn Health) { /* ... */ }
+///
+/// fn no_inference(i: &mut dyn NoInference) { /* ... */ }
+///
+/// // Type can be inferred since `static supertrait is explicitly declared for Health trait.
+/// let mut dyn_gd = Monster::new_gd().into_dyn();
+/// deal_damage(&mut *dyn_gd.dyn_bind_mut());
+///
+/// // Otherwise type can't be properly inferred.
+/// let mut dyn_gd = Monster::new_gd().into_dyn::<dyn NoInference>();
+/// no_inference(&mut *dyn_gd.dyn_bind_mut());
+///
+/// // Type can't be inferred.
+/// // Would result in confusing compilation error
+/// // since compiler would try to enforce 'static **lifetime** on our reference.
+/// // let mut dyn_gd = Monster::new_gd().into_dyn();
+/// // no_inference(&mut *dyn_gd.dyn_bind_mut());
```

Comment:
This could be a separate block with [`compile_fail`](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#attributes).

Reply:
Also, instead of `**lifetime**`, you could write `*lifetime* (&'static mut ...)`.